### ISSUE TITLE
ConvertTo-Json to auto detect depth, or print related warning, PLEASE.

### ISSUE THREAD

**Issue by Garfius** [PowerShell/PowerShell#23960]
### Summary of the new feature / enhancement

I was just constructing bodies to be used via REST, and lost an hour to find out ConvertTo-Json needs -depth specified🙄.

So, please get depth auto-detection or at least print a warning.

Actual Warning status
- Windows: No Warning, nor error😵.
- Linux: already prints the Warning.

The message on Linux:
`WARNING: Resulting JSON is truncated as serialization has exceeded the set depth of 2.`

Test code:
`@{"firstKey"="value1";"secondKey"= @{"nested1Key"=@{"nested2Key"=@("one","two")}}} | ConvertTo-Json`

**Comment by iRon7** [comment]
Which PowerShell version are you testing?

Knowing that [PowerShell (core)](https://learn.microsoft.com/powershell/scripting/install/installing-powershell) has already undergone several enhancements towards the default [`ConvertTo-Json -depth`](https://learn.microsoft.com/powershell/module/microsoft.powershell.utility/convertto-json#-depth) parameter and should in fact react the same as the version under Linux:

<img width="681" alt="image" src="https://github.com/PowerShell/PowerShell/assets/8080834/6d62237c-9d1e-4fb0-95cc-cb4fa1f99c27">

### Windows PowerShell
Windows PowerShell issues or suggestions, might be opened [here](https://support.microsoft.com/windows/send-feedback-to-microsoft-with-the-feedback-hub-app-f59187f8-8739-22d6-ba93-f66612949332)
<sup>(but from my experience with this site, I give it little change that ***Windows*** PowerShell will change for this)</sup>

**Comment by 237dmitry** [comment]
> already undergone several enhancements

It shows warning but does not automatically set desired depth

**Comment by iRon7** [comment]
@237dmitry, [Some objects have circular references, meaning that a child object could refer to a parent (or one of its grandparents) causing a infinitive loop if it would be serialized to JSON.](https://stackoverflow.com/a/53583678/1701026)

**Comment by 237dmitry** [comment]
The depth indication is not intuitive for the user. In fact, this is a real problem; there is a high probability of corrupting the resulting json file. Previously, when I used Windows Terminal, sometimes I configured it via settings.json, a couple of times I forgot to specify the depth, and was forced to replace it with a saved copy. It would be better to set the default depth to 10 or 100 instead of 2.

**Comment by iRon7** [comment]
@237dmitry,

> 
It would be better to set the default depth to 10 or 100 instead of 2

This is an old discussion, see the dupplicate: #8393

**Comment by 237dmitry** [comment]
> This is an old discussion

Has anything changed in six years?

**Comment by RG255** [comment]
This is an issue in other places as well as I found out the hard way. I have a Named Pipe set up and I use a PSObject to pass the data back and forward from client to server and vice versa using:

`$PipeInfo.Writer.Writeline([Management.Automation.PSSerializer]::Serialize($DataObject,[int]1) -replace ('([\r]|[\n]|[\t])') -replace ('>[ ]+<', '><'))`

The -replace regex removes all the extraneous CR, LF and TAB characters and then the spaces between the ">" and "<" characters in the resulting XML to reduce the size. At the other end of the pipe, I have:

`$result=[Management.Automation.PSSerializer]::Deserialize($PipeInfo.Reader.ReadLine())`

The depth issue can be varied as the objects returned depend on the results of the actions carried out either by the data sent by the client or returned by the server.

If there is no warning and no straightforward way to work out the depth - what are my options?

What value should I make the [int]1?

**Comment by mklement0** [comment]
@RG255, this is a separate issue, as it pertains to PowerShell's _general-purpose serializer_, as used in remoting and background jobs, which:

* defaults to depth `1` for non-DTOs
* has _no_ depth limit on DTOs.

DTOs (data-transfer objects) in the context of PowerShell are (nested) hashtable (dictionary) or `[pscustomobject]` instances.

As for what depth you should pass for non-DTOs in custom-serialization scenarios: that depends on the specify types you're serializing, but note that even relatively low numbers can quickly get out of hand - that's why the default is so low.
Non-DTOs are by definition not designed with serialization in mind, so it makes sense to depth-limit them to prevent "runaway serialization".

Applying this distinction _analogously_ to `ConvertTo-Json` was previously proposed in detail in the following comment:

* https://github.com/PowerShell/PowerShell/issues/8393#issuecomment-445656280

This would:
* fully serialize DTOs, with (virtually) no depth limitation (except, possibly, if an _explicit_ `-Depth` argument is given).
* impose the default depth of `2`  on non-DTO objects _only_

In other words:
 * with DTOs, it would just work, without the need for a warning
 * with non-DTOs - which aren't typically passed to `ConvertTo-Json` - truncation at the default / specified depth will occur; whether the current warning is then even needed is debatable.
